Add chat history accessible with up/down arrows

Argmaster opened the pull request:
This pull request adds mechanism widely present in terminals which allows user to quickly access commands they typed before by using up and down arrows. Implementation has few assumptions:

- When editing multiline message previous/next command should be accessed only from first and last line (respectively)
- When modifying command/message from history original history entry has to be preserved
- When moving "up" history it should be always possible to return "down" even if modified message
- Modified messages are preserved when moving across history entries
- Duplicated messages are not preserved

archbirdplus commented on src/gui/windows/chat.zig:
```diff
@@ -87,17 +102,57 @@ fn refresh() void {
+	}
+
+	input.clear();
+	for(msg) |c| input.inputCharacter(c);
```
It might be nice to have a bulk "append string" method in `TextInput`, just like `clear` and `paste`.

Argmaster replied:
Yeah, maybe, I'll do that, but on the other hand its first time it is done and performance of current solution is not particularly bad, so I didn't find It necessary.

Argmaster commented on src/gui/windows/chat.zig:
```diff
@@ -87,17 +102,57 @@ fn refresh() void {
+	var msg = inQueue.dequeue_front().?;
+	if(std.mem.eql(u8, msg, current)) {
+		main.globalAllocator.free(msg);
+		if(inQueue.dequeue_front()) |m| { msg = m; }
```
It should **also** happen while appending the messages.

Argmaster commented on src/gui/windows/chat.zig:
```diff
@@ -37,9 +40,13 @@ var historyStart: u32 = 0;
 pub var input: *TextInput = undefined;
 var hideInput: bool = true;
+var upHistory: CircularBufferQueue([]const u8) = undefined;
+var downHistory: CircularBufferQueue([]const u8) = undefined;
```
There is more subtle gain from it compared to single queue history. Assume you have history and move up twice to a command, then you modify it and by mistake move up again. With current system modified message will be inserted in the middle, between history entry you were modifying and history entry you moved to, so you can always go back.

So other options either don't preserve entries you modify or preserving modified entries is O(N).

If you think otherwise please provide more details, I am no data structure specialist, I do stuff so it works the way I want it rather than making it optimal from the beginning.

archbirdplus replied:
Oh, your model here is totally different from what I expected. I expected that you just had a list of old messages, and you could modify each one (or all) as you wished, and that scratch history would get replaced as soon as you entered something. Coincidentally like my favorite `zsh`.

What you have is that you can scroll through each message, and edit them, but when scrolling away from an edited message you first insert it into the history list. I mean, if that's intended behavior, your data structure is pretty good (though do you need to duplicate the string before immediately freeing it?). I get it now, but the user experience is quite surprising.

Oddities:
1) The first time you press up or down after editing a message, the message box "undoes" your message rather than actually scrolling to the message before/after the one you were just editing.
2) Message boxes get cloned every time you scroll up or down, and this can get quite cluttered if you move back and forth. I don't think there's a great reason to preserve the edit-histories of history like that. Generally this would be written into an invisible undo-buffer to be accessed with <Meta key>+Z.
2) Edits to messages get preserved after pressing enter. Combined with the previous issue, it means that the intermediate edits linger up in your history, while the final drafts ironically get removed.

Overall, I think this is way too conservative. I don't think chat messages are precious enough that every change must be remembered. Most software (besides `vim`, which has 2 dimensions of undo) would start dropping edits much earlier than this.

I just checked, and it turns out that shells also have different ways of dealing with history. My `zsh` keeps history edits when scrolling, then clears them when any command is run. Apparently my `bash` keeps history edits _until that specific line_ is entered, and only resets one line at a time. So there is variation there.

Now I see what you're getting at, but I also think you're just trying to serialize a list of undo histories (a 2d array) into the main edit history. So I guess I would implement this as an array of doubly-linked lists (rather than your center-insertable circular buffer, which I do think is quite clever). This feels like a natural modern GUI generalization of shell-style edit histories. Thus you could scroll through the history above, have it not desync from the displayed history, and you can edit it at your leisure, and your can use standard UI conventions to recover from accidents.

Wow this is quite a long message, but there are a surprising number of subtleties going on here. Basically, your UI is rather idiosyncratic, which wouldn't help your PR or first-time users.

Argmaster replied:
Where exactly do I dupe message just before freeing it?

archbirdplus replied:
I think there is a balance between "put the entire edit tree into the history" and "remove all changes when you look away". This balance for me is to allow whatever temporary edits you want in the history, to be cleared entirely but only when you press enter.

Do you really go through multiple divergent iterations of each command in the editor before sending any of them? Do you need to see all of your previous unsent iterations? If so, this model works fine. But it is more afraid of losing work than anything I've seen before. It is idiosyncratic because these are fresh rules that users will be seeing for the first time.

I think oddity 3 is arguable as a feature, so I won't debate that. However, oddity 1 feels distinctly unintentional.

Argmaster replied:
I just checked in minecraft, and it has most of the behavior described here - it does preserve old messages, but commits new message to history only after enter is hit. I does have 1. behavior, ie. moving one up restores history entry to unmodified state, but modified versions are lost when you navigate away from them.

Powershell preserves modified versions of commands until you commit one of them (in between history entries), then only that one you have committed is preserved. When you go through history and modify one of the entries it appears that history pointer is reset, so it doesn't exactly have 1. behavior, unless you were at the beginning of the history, then naturally next entry is unmodified version of your current state.

Argmaster replied:
Overall, I think there are so many different implementations of history behavior that none of them can be called idiosyncratic unless all of them can be called that.

Argmaster replied:
IMO PowerShell behavior would be most optimal, but it would require rewriting most of the code for very small (and ambiguous) improvement of UX compared to having history at all. I would prefer to settle on what we currently have, if you want you can always make it better later on, for now I would appreciate having anything.

archbirdplus replied:
I am just not sure how to interpret your text here. Just to clarify:
Minecraft deletes any edits to history when you scroll away, only appending edits when sent?
Powershell keeps modifications to all messages in history, but when sending a message it is appended to history, and only the edited message that got sent is reset to its initial value?

As for oddity/behavior 1, I mean something different. It took me a long time to figure this out:
Consider the message history:
```
A
B
C
```
Scroll to message B, change it to say D, then scroll up or down. Does this create
```
A
D
B
C
```
or
```
A
B
D
C
```
I hope I am not tripping. This behavior depends on whether you were at A before editing B, or if you were at C before editing B. It seems history gets put in between where you last were before, vs during editing. This is very confusing. It is also not what you described previously. Please look into this.

Additionally you cannot scroll down past C to get to the empty message again.

archbirdplus replied:
If I read correctly, then Powershell is using `bash` behavior. Minecraft is just being lossy. Nobody inserts multiple edits back into history. But I am not strictly opposed to it if it makes enough sense.

Argmaster replied:
bash (at least git-bash I have access to rn) behaves differently than powershell, it allows you to modify the history and it does not preserve original history entries after you modify them.

Argmaster replied:
Overall, bash behavior preserves least information, but its easiest to implement, powershell behavior is most balanced between preserving and keeping information at once, but seems most complicated to do, current implementation from this PR preserves the most information and IMO implementation is quite simple and there is also minecraft which just sucks.

archbirdplus replied:
That is an incredible diversity. Perhaps worth writing about 😄. Again, not sure about the meaning of the last paragraph ("If I observe current behavior correctly") but I think it is worth making this behavior more stable. e.g., always put the edited message below its parent, then move 1 spot up or down from the edited message based on the key. That way there's a trend of edited messages being closer to the bottom rather than dispersing all around the original.

IntegratedQuantum replied:
I also think it would be nicer if the message was inserted below the original and not above it. It also shouldn't be too much trouble to implement, right?

Argmaster replied:
Honestly, I can't think of simple way to handle that, I have no way of tracking if message is older or newer when I move it between histories, your request would require code to do that, otherwise there is no way to distinguish between new message, modified history message and unmodified one.

IntegratedQuantum replied:
Well, instead of putting it above the command that's modified, you could just put it below it???

Argmaster replied:
Ok, I kind of figured that out, so now behavior is as follows:

---

type: `A <Enter>`
type: `B <Enter>`
type: `C <Enter>`

History:

```
A
B
C
```

textInput: ``

type: `<upArrow>`

textInput: `C`

`<upArrow>`

textInput: `B`

type: `12`
textInput: `B12`

---
And now from that state:

## 1. Scenario 

`<upArrow>`

textInput: `B`

History:

```
A
B     <-- we are here
B12
C
```

## 2. Scenario

`<downArrow>`

textInput: `C`

```
A
B
B12
C     <-- we are here
```

Rest should be more or less the same. I would appreciate if @archbirdplus and @IntegratedQuantum could try to break it cause I have no faith that it will work correctly.

IntegratedQuantum replied:
There seems to be an issue when changing the last history entry:
If you press up after changing the oldest entry, it will still not navigate to the previous value of it.
But to be honest, with an 8192 history buffer and since this will likely be saved in the future too, I don't think it matters much.
So we can keep it bugged if that keeps the implementation simpler.

IntegratedQuantum commented on src/gui/components/TextInput.zig:
```diff
@@ -393,6 +412,16 @@ pub fn inputCharacter(self: *TextInput, character: u21) void {
 	}
 }
 
+pub fn inputString(self: *TextInput, utf8EncodedString: []const u8) void {
```
The functionality you actually need is `setString`, why not implement that function instead?

IntegratedQuantum replied:
setString would also clear it, making the explicit clearing at the callsite unnecessary.

IntegratedQuantum commented on src/gui/windows/chat.zig:
```diff
@@ -1,6 +1,7 @@
 const std = @import("std");
 
 const main = @import("root");
+// const main = @import("../../main.zig");
```
unintended change

IntegratedQuantum commented on src/gui/windows/chat.zig:
```diff
@@ -29,6 +31,7 @@ pub var window: GuiWindow = GuiWindow{
 const padding: f32 = 8;
 const messageTimeout: i32 = 10000;
 const messageFade = 1000;
+const reusableHistoryMaxSize = 128;
```
This memory costs nothing, but if a user does hit the "end of history", then it is rather painful (trust me I've been there)
I'd at least set it to 10000 or something.

Argmaster replied:
Increased to 8192, I like multiples of 2

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -362,6 +362,10 @@ pub fn CircularBufferQueue(comptime T: type) type { // MARK: CircularBufferQueue
 			}
 		}
 
+		pub inline fn enqueue_front(self: *Self, elem: T) void {
```
I agree that the current variant is not quite optimal. But instead of aliasing it I would prefer to only have one way to do things. (of course this should be done in a separate PR, and it's fine to keep it like this here, though I would appreciate making an issue or adding a TODO note here for future reference)

Argmaster replied:
#1320

Argmaster commented on src/gui/windows/chat.zig:
```diff
@@ -37,9 +39,108 @@ var historyStart: u32 = 0;
+	}
+	fn moveUp(self: *History) void {
+		std.debug.assert(self.canMoveUp());
+		// This can not overflow because we are moving items between two queues of same size.
```
I think what I ment was that down queue can at most reach its capacity when moving items from up queue to down queue. Howeve when I think about it right now it is not necessarily true since modified messages are preserved, so if up queue is full, we move half to down queue, modify message, move up and then move rest of the messages to down queue it will overflow by exactly 1 message. We should have defined behavior in such case. Now it would result in queue reallocating which is not intended behavior.

IntegratedQuantum replied:
How about using the new fixed size queue, added with the network system?

Argmaster replied:
Yes, we should do that.

Argmaster commented on src/gui/windows/chat.zig:
```diff
@@ -37,9 +39,108 @@ var historyStart: u32 = 0;
+		return self.up.empty() == false;
+	}
+	fn moveUp(self: *History) void {
+		std.debug.assert(self.canMoveUp());
```
This is redundant because you have looked up the implementation of those methods. Those assertions are complementary to guarantee what you have just checked in code.

Argmaster commented:
I have changed the code to use FixedSizeCircularBuffer.
Now there is a bit confusing edge case that if `up` queue is full and you move all the way up so all the messages are in `down` queue and after that you modify one of the messages, the latest message in `down` buffer will be discarded, since `forceEnqueueFront` does `dequeueBack` if queue is full. However considering the history has 8k entries, it is really unlikely people will navigate all the way up. There is no simple workaround, as player can generate infinitely many modified messages, It would require some kind of special case and imo it is not worth the effort.

Argmaster commented:
Alright I also made it so it is possible to return to empty entry if you have navigated up, but it is not saved.

Argmaster commented on src/gui/windows/chat.zig:
```diff
@@ -37,9 +39,108 @@ var historyStart: u32 = 0;
+	}
+	fn moveDown(self: *History) void {
+		std.debug.assert(self.canMoveDown());
+		// This can not overflow because we are moving items between two queues of same size.
```
Alright, I have that comment because it was simply not true.

Argmaster commented:
Alr, I think I have addressed comments from previous review.

IntegratedQuantum reviewed (changes requested):
There seem to be some problems in behavior:
Example one from the start doing:
`a` `Enter` `↑` `ab` `↑`
After this sequence I expect to be at `a`, since `ab` is inserted after `a`. However my location is still at `ab`. If I press `↓` `↑` `↑` I correctly get to `a`.

Second example:
`a` `Enter` `b` `Enter` `↑` `bc` `↑`
Since the modified string should be below `b`, I expect to be at `b`, however this sequence gets me to `a`.
Going down I first get to `b` then to `bc`, which confirms that the order in the buffer is correct.

Argmaster commented:
That is an unfortunate edge case I can't really get rid off, I tried adding `loadPreviousHistoryEntry` if `!messageHistory.canMoveUp()` but to no avail, I will try some different options tomorrow. It happens only on last history entry I think so it is not that visible.

Argmaster commented:
Alright, I think I figured it out!
I suspect that there might be some strange behaviors here and there with deduplication logic, but the core functionality should be there with jumping to history entries behaving like it should...

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -408,6 +459,15 @@ pub fn FixedSizeCircularBuffer(T: type, capacity: comptime_int) type { // MARK:
 		}
+
+		pub fn isEmpty(self: Self) bool {
+			return self.len == 0;
```
Do we really need to abstract this? In my opinion it is nicer to directly see what it's doing instead of having an opaque method for things like this. If used externally you might have a point, but it is only used internally.

IntegratedQuantum replied:
> Opaque methods are there so you don't need to think about the implementation details after you implement that thing once.

I would generally agree, but here we are in the implementation detail.
And if I do want to read or debug the implementation details of a data structure then I don't want to have to step through 10 different messages just to figure out what's actually going on.

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -408,6 +459,15 @@ pub fn FixedSizeCircularBuffer(T: type, capacity: comptime_int) type { // MARK:
+			return self.len == 0;
+		}
+
+		pub fn isFull(self: Self) bool {
```
This is an internal helper function and should be private.
I can't think of any use case for using this externally.

IntegratedQuantum replied:
Also maybe the same thing applies as for isEmpty:
Why not just keep this in the places where it's called?
E.g. functions like `insertSlice` also explicitly check against the capacity.

IntegratedQuantum commented:
behavior is good now from my perspective, unless @archbirdplus finds any more issues

archbirdplus commented:
I did not get pinged 🤔. Anyways:

If I start with a blank message, and I scroll up, I should be able to return to blank message when I scroll back down. Think of it as peeking up at my last message. Currently the last message gets duplicated instead. That's the main issue.
I know you currently remove history lines that are empty, but at least special-case the last one. I would treat this as an empty line that is accessible until you try to write something in it. You could also just create an empty line each time you scroll to the bottom, if that is easier.

Next, it seems like if you have the history:
```
a
b
c
```
Scrolling to `b`, replacing it with `a`, then scrolling up, will not change history. On the other hand, scrolling to `a`, replacing it with `b`, and scrolling down will duplicate `b` permanently, which is not symmetric. I believe this happens regardless of the direction you scroll away to.
It might be two problems under the hood. The first is that `a` below `b` is somehow de-duplicated, even though they're not adjacent. The second is that `b` below `a` is not de-duplicated, even though it is adjacent. These two are secondary issues though.

Argmaster commented:
Having an empty message anywhere is unfortunately not possible, at least not in a trivial way. If you insert an empty message at the beginning it will either get removed by reduplication mechanism or it will produce more of them in least expected places. I did try that but with no luck. However, you can still remove any message without losing that message which is a viable compromise in my opinion. If you want you can try to improve this system without losing its major advantage over scratch history - never losing unique messages.
As for the lack of symmetry in deduplication, I will see what I can do about it. But that is again the emerging behavior of this system.

archbirdplus commented:
I accept the first compromise in all places but one. Consider opening the chat. Above you have
```
a
b
c
```
The current textbox is empty. Scroll up, then scroll down. The current textbox does not return to empty. This is the one empty message I want to keep. Instead, starting from the blank message, `Up` `Down` ends up with `c`, and pressing `Up` is _still_ `c`, which means the `c` got duplicated for some reason. I do like the behavior of removing empty messages in the middle of history (this lets you remove bad uncommitted changes), but the bottommost empty message should be special-cased.

As for the emergent behavior, I think it might have to be modified to match the intuition closer. In this case, maybe it needs to coalesce neighboring messages in some common code before making direction-dependent changes.

IntegratedQuantum reviewed (changes requested):
I also was bothered by this a bit, so it's probably worth figuring out how to fix it.

Also you don't actually need an empty string anywhere, you can just implement this as a special case when down.

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -338,12 +338,53 @@ pub fn FixedSizeCircularBuffer(T: type, capacity: comptime_int) type { // MARK:
+		}
+
 		pub fn enqueue(self: *Self, elem: T) !void {
-			if(self.len >= capacity) return error.OutOfMemory;
```
I hate to be annoying here, but honestly I kind of prefer keeping it as `>=`

IntegratedQuantum replied:
Yeah, I know, I just changed my mind

Argmaster commented:
@archbirdplus  try it now.

archbirdplus commented:
Nice. Now, in order of priority:
- Editing the blank message keeps it in history. I would rather you just overwrite the blank, but continue storing all revisions after that.
- The most bottommost message is still duplicated. By this I mean the very last message that hasn't been sent yet, and which you find yourself on when you first open chat. If you open chat (or otherwise look at this bottommost message), then scroll down, then scroll up, you still have the same message. This makes it look like it has been duplicated, even though it gets fixed when you add another line after it. This is what I described in my last comment.
- Scrolling up and hitting enter to resend a previous message doesn't add it to history. Not sure if this is intended, but I wouldn't block the PR over this.

And on top of that the de-duplication bug still persists.

Argmaster commented:
@archbirdplus alr so, I am too stupid to fully understand what you are saying, but I changed it again a bit, and I cant really reproduce the duplicated message behavior so for me it seems alr... Could you check again and pretty please provide simple step-by-step guide how to reproduce any undesired behaviors you find?
Not saving messages that already appear in history if they were selected from history was intentional but I changed it bcs it was counterintuitive. Blank entries should not appear in history, but blank entry at the beggining of the history is immortal due to how it is implemented and will stay there, no other blank entries will stay in history (like 99% sure).

archbirdplus commented:
Sorry.
Issue 1:
- Open chat. Press `a` `Enter` `b` `Enter` `c` `Enter`. Current textbox: ` `. Press `d`. Current textbox: `d`. Press `↑`. Current textbox: ` `. Instead, it should be `c`. Press `↓` `↓`. This can either be `d` or ` `, up to you, but the blank higher up should be removed.

Issue 2:
- Open chat. Press `a` `Enter` `b` `Enter` `c` `Enter`. Current textbox: ` `. Press `↓` `↑`. Current textbox: ` `. Instead, it should be `c`. The ` ` appears to have been duplicated and you can scroll into it twice.

Issue 4:
- Open chat. Press `a` `Enter` `b` `Enter` `c` `Enter`. Current textbox: ` `. Press `↑`. Current textbox: `c`. Press `Delete` `b`. Current textbox: `b`. Press `↑`. Current textbox: `b`. Instead, it should be `c`. The history is `a` `b` `c` ` `. Instead, the history should go `a` `b` `c` `b` ` ` because edits should go below their parents.

- Open chat. Press `a` `Enter` `b` `Enter` `c` `Enter`. Current textbox: ` `. Press `↑` `↑` `↑`. Current textbox: `a`. Press `Delete` `b`. Current textbox: `b`. Press `↓`. Current textbox: `b`. Although its a bit ambiguous, it should probably be `c` because you de-duplicate history. Currently the history reads `a` `b` `b` `c` ` `. Instead, the history should go `a` `b` `c` ` ` and one of the `b` should be trimmed.

These are the setups that I am testing. Tell me if you disagree or  can't replicate something.

Argmaster commented:
Issue 1 - can't reproduce after last change.
Issue 2 - can't reproduce after last change.
Issue 4 a and b - The fact that you have manually typed contents of history entry doesn't make your message a history entry. Your message was not yet part of history, only after you moved down/up, it became part of history and since it was a duplicate it was removed. Harmless and arguably preferable approach. Jumping two positions if your message was a duplicate would be more surprising IMO. It should also be **very** rare in average use.

archbirdplus commented:
Issue 1 is resolved. Issue 2 is resolved. Ok, great!
I agree that issue 4 is very rare, but it has the appearance of being an intentional feature, although slightly broken.

For 4a:
What I dont understand is why editing `c` to be `b` would make it de-duplicated. Isn't the logic supposed to be that edits come _after_ the original message, therefore suggesting `a` `b` `c` `b` ` ` instead?

For 4b:
> Jumping two positions if your message was a duplicate would be more surprising IMO

Alright, I'll take that. Then the 4b problem is slightly more reasonable, but there are still 2 consecutive `b` entries in history. They weren't removed. And if this message isn't removed, then I don't understand why the message in 4a _is_ removed.

archbirdplus commented:
I thought I saw it again, but I can't reproduce 4. But I found one more- when you scroll up, exit chat, then press `/`, the chat gets replaced with `/` but you stay in the same place in history. I think its fine to keep the location of history after exiting, but not with the `/` shortcut. That should be all.

Argmaster commented:
@archbirdplus Ehh that is a bad one, I already made tools to fix this in #1380 (onCharacter callback) but fixing conflicts caused by adding it here would suck, can we note this down as known limitation and fix this after #1380? Basically fix would be to flush history in onCharacter if all input contains is '/', just adding that callback now is not necessarily fun experience C:
I suppose I should ask @IntegratedQuantum the same C:

archbirdplus commented:
As long as the bug won't be left hanging indefinitely, I approve this PR from the functionality perspective.

Argmaster commented:
I will make PR for that today

IntegratedQuantum commented on src/utils.zig:
```diff
@@ -376,6 +417,10 @@ pub fn FixedSizeCircularBuffer(T: type, capacity: comptime_int) type { // MARK:
 		}
 
+		pub fn dequeueFront(self: *Self) ?T {
+			return self.dequeue();
```
There is a mismatch in the naming conventions.
Below, in `CircularBufferQueue`, you delcared `dequeue_back` as `self.dequeue`

Argmaster replied:
Yes, but I suppose we prefer `camelCase` rigth? That one should be gone + it was done that way because of other methods being `snake_case`

Argmaster replied:
I have removed changes from `CircularBufferQueue`

IntegratedQuantum replied:
It's also about back vs front naming

Argmaster replied:
Hmm, right, I'll swap those

IntegratedQuantum commented on src/items.zig:
```diff
@@ -463,7 +463,7 @@ pub const Tool = struct { // MARK: Tool
 		// TODO: This is leaking textures!
 		//if(self.texture) |texture| {
-			//texture.deinit();
+		//texture.deinit();
```
What is this doing here?

Argmaster replied:
I have no idea why it still shows in diff, if the master version is identical

Argmaster replied:
Alr now its gone!

IntegratedQuantum reviewed (approved):
Now I'm still confused on the naming convention, I looked it up on [wikipedia](https://en.wikipedia.org/wiki/Double-ended_queue#Operations) and consensus seems to be that `back` is at the end of the range, which kind of makes more sense.

But I don't want to keep this PR blocked on naming issues any longer, so please resolve this as a part of #1320, and please do it sooner rather than later.
